File: 2024-1/week_2/김민규/budget.js
// 입력 처리
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let input = [];
rl.on('line', function (line) {
    input.push(line);
}).on('close', function () {
    // 입력 처리
    const N = parseInt(input[0]); // 지방의 수
    const budgets = input[1].split(' ').map(Number); // 각 지방의 예산 요청
    const M = parseInt(input[2]); // 총 예산

    // 여기서부터 문제 해결 로직을 작성하면 됩니다.

    // 출력 (최댓값을 구한 후 출력)
    const answer = solve(N, budgets, M);
    console.log(answer);
});

function solve(N, budgets, M) {
    budgets.sort((a, b) => a - b);
    const total = budgets.reduce((prev,cur)=> prev+cur)
    if(total <= M) return Math.max(...budgets)

    let current = 0
    let dum = M;
    let result = 0;

    for(let i = 0; i< N; i++){
        let currentBudget = budgets[i];
        let dumCity = N -i;

        if(dum >= currentBudget * dumCity){
            dum -= currentBudget
            result = currentBudget
        }else{
            result = Math.floor(dum / dumCity);
            break
        }
    }
    return result

}

File: 2024-1/week_2/김민규/budget.py
def solve(N, budgets, M):
    budgets.sort()  # 예산 요청을 정렬
    total = sum(budgets)  # 예산 요청의 총합 계산
    if total <= M:
        return max(budgets)  # 총합이 예산 내라면, 최대값 반환

    dum = M
    result = 0

    for i in range(N):
        current_budget = budgets[i]
        dum_city = N - i

        if dum >= current_budget * dum_city:
            dum -= current_budget
            result = current_budget
        else:
            result = dum // dum_city  # 남은 예산을 나누어줌
            break

    return result

# 입력 받는 부분
N = int(input())  # 지방의 수
budgets = list(map(int, input().split()))  # 각 지방의 예산 요청
M = int(input())  # 총 예산

# 해결 및 출력
answer = solve(N, budgets, M)
print(answer)